File: benchmark/fusion/deforestation/page7-1.f
Inductive List = nil Unit | cons {Int, List};

map = \f: Int -> Int. fix (
  \g: List -> List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> cons {f h, g t}
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

square = \x: Int. * x x;

upto = \n: Int. (fix (
  \f: Int -> Compress List. \now: Int.
  if > now n then nil unit
  else cons {now, f (+ now 1)}
)) 1;

/*sqsum = \n: Int. / (* n (* (+ n 1) (+ n (+ n 1)))) 6;*/

main = \n: Int. sum (map square (upto n));

File: ml-benchmark/fusion/deforestation/page7-1.ml
type l = Nil | Cons of int * l

let map f =
  let rec g xs =
    match xs with
    | Nil -> Nil
    | Cons (h, t) -> Cons (f h, g t)
  in
  g

let rec sum xs =
  match xs with
  | Nil -> 0
  | Cons (h, t) -> h + sum t

let square x = x * x

let upto n =
  let rec f now =
    if now > n then Nil
    else Cons (now, f (now + 1))
  in
  f 1

let program n = sum (map square (upto n))